File: Data Structures/Fenwick Tree/Fenwick Tree.py
class FenwickTree:
    neutral = 0  # define according to the problem

    def __init__(self, N):
        self.ft = [self.neutral] * (N + 1)

    def LSOne(self, i):
        return i & (-i)

    # define according to the problem
    def op(self, x, y):
        return x + y

    # define according to the problem
    def opInv(self, x, y):
        return x - y

    def rangeQuery(self, low, high):
        return self.opInv(
            self.rangeQueryI(high + 1),
            self.neutral if low == 0 else self.rangeQueryI(low),
        )

    def rangeQueryI(self, i):
        acc = self.neutral
        while i > 0:
            acc = self.op(acc, self.ft[i])
            i -= self.LSOne(i)
        return acc

    def update(self, i, val):
        prev = self.rangeQuery(i, i)
        i += 1
        while i < len(self.ft):
            self.ft[i] = self.opInv(self.ft[i], prev)
            self.ft[i] = self.op(self.ft[i], val)
            i += self.LSOne(i)

File: Data Structures/Fenwick Tree/Fenwick Tree.cpp
template <class T>
class FenwickTree
{
private:
    T neutral = 0; // define according to the problem
    vector<T> ft;

    int LSOne(int i)
    {
        return i & (-i);
    }

    // define according to the problem
    T op(T x, T y)
    {
        return x + y;
    }

    // define according to the problem
    T opInv(T x, T y)
    {
        return x - y;
    }

public:
    FenwickTree(int N)
    {
        ft.assign(N + 1, neutral);
    }

    T rangeQuery(int i)
    {
        T acc = neutral;
        for (; i > 0; i -= LSOne(i))
            acc = op(acc, ft[i]);
        return acc;
    }

    T rangeQuery(int low, int high)
    {
        return opInv(rangeQuery(high + 1), (low == 0 ? neutral : rangeQuery(low)));
    }

    void update(int i, T val)
    {
        T prev = rangeQuery(i, i++);
        for (; i < ft.size(); i += LSOne(i))
        {
            ft[i] = opInv(ft[i], prev);
            ft[i] = op(ft[i], val);
        }
    }
};